README.md
# treeseg

Extract individual trees from lidar point clouds

<img src="https://4xnnka.bl3302.livefilestore.com/y4mzpqPKgclbQKR51t4a9tA2v37IKMp221dDuLXHvRqQzc8szMleP7anmHxoy-bPwa-0qnDRrRpQkBhkO2P3cef-u7_eBpNubE_wO8FA5E0sDkbmPt0BIJoWT2upX_YFO5UfTS0ys0LZLHaSl0DktwdJIj82R4ZP051-2CiPzJAXJvht3eijlwfp0mo_W8VeXVa8p_g79_EEb51Kg0WGMWe8A?width=660&height=628&cropmode=none" width="500">

## Overview

*treeseg* has been developed to near-automatically extract tree-level point clouds from high-density larger-area lidar point clouds acquired in forested scenes.
A full description of the methods can be found here: http://discovery.ucl.ac.uk/1575534/ (pp. 96-167)

Briefly, libtreeseg provides a set of generic functions that have been wrapped here into a series of executables to: i) identify individual stems in the larger-area point cloud (findstems), ii) extract each of these identified stems up to the first order of branching (segmentstem), and iii) segment each individual crown from neighbouring vegetation (segmentcrown).

Integrated into *treeseg* is libleafsep, a library providing methods to separate wood and leaf returns in unorganised point clouds. 
libleafsep is a modified translation of TLSeparation (https://github.com/TLSeparation).

## Prerequisites

*treeseg* has been developed using:

Point Cloud Library (PCL) (http://pointclouds.org)
Armadillo (http://arma.sourceforge.net)

On RHEL 7, these dependencies can be installed via:

```
yum install cmake
yum install pcl pcl-devel pcl-tools pcl-doc
yum install armadillo armadillo-devel
```

On macOS 10.13, dependencies are perhaps most easily installed via *Homebrew* (https://brew.sh):

```
brew install cmake
brew install pcl
brew install armadillo
```

## Installation

Install libtreeseg and the associated binaries as follows:

```
git clone https://github.com/apburt/treeseg.git;
cd treeseg;
mkdir build;
cd build;
cmake ../src;
make;
```

Also included in *treeseg* is rxp2pcd, for conversion of REIGL V-Line scan data to .pcd binary format. Linux-only, this requires the *RiVLIB* headers and libraries (downloadable from: http://www.riegl.com/index.php?id=224) to be installed in /treeseg/include/riegl/ and /treeseg/lib/ respectively.

## Usage

Below is an example usage of the *treeseg* binaries:

* plotcoords ../matrix/ > nouragesH20_coords.dat
* rxp2pcd ../ nouraguesH20_coords.dat 60 15 nouraguesH20
* nearestneighbour 1 4 nouraguesH20.sample.pcd > nouraguesH20_nn.dat
* downsample 0.04 1 nouraguesH20_*.pcd
* getdemslice 2 3 6 nouraguesH20_*.downsample.pcd
* findstems 15 0.2 2 ../nouraguesH20_coords.dat ../nouraguesH20.slice.downsample.pcd
* segmentstem 12.5 ../nouraguesH20.downsample.pcd ../clusters/cluster_*.pcd
* getcrownvolume ../nouraguesH20.downsample.pcd ../stems/stem_*.pcd
* segmentcrown [14 - 16] ../volume_*.pcd

## Authors

* **Andrew Burt**
* **Mathias Disney**
* **Kim Calders**
* **Matheus Boni Vicari**

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

## Acknowledgements

* *treeseg* makes use of the Point Cloud Library (PCL) (http://pointclouds.org) and Armadillo (http://arma.sourceforge.net).
* AB was supported by NERC CASE PhD NE/J016926/1 with EADS Astrium; additional funding was via the NERC National Centre for Earth Observation (NCEO), NERC funding awards NE/N00373X/1, NE/P011780/1 and NE/K002554/1; MD is also supported by the EU Horizon2020 project (BACI project funded by the EU's Horizon 2020 Research and Innovation Programme under grant agreement 640176).


CHANGES
1.0.0 2017-12-10
	Initial release
1.0.1 2018-01-08
	Bug fixes
1.1.1 2018-03-10
	Added libleafsep for separating wood and leaf returns
	segmentcrown performance fix


src/rxp2pcd.cpp
//Andrew Burt - a.burt@ucl.ac.uk

#include <string>
#include <sstream>
#include <fstream>
#include <cstdlib>
#include <vector>
#include <cmath>

#include <dirent.h>

#include <riegl/scanlib.hpp>

struct pcloud 
{
	std::vector<float> x,y,z;
	std::vector<float> range;
	std::vector<float> amplitude;
	std::vector<float> reflectance;
	std::vector<float> deviation;
	std::vector<float> return_number;
	float scan_number;
	std::vector<double> time;
	float matrix[16];
};

class importer : public scanlib::pointcloud
{	
	pcloud &pc;
	public:
		importer(pcloud &pc) : scanlib::pointcloud(false), pc(pc){}
	protected:
		void on_shot_end()
		{
			for(int i=0;i<targets.size();i++)
			{
				pc.x.push_back(targets[i].vertex[0]);
				pc.y.push_back(targets[i].vertex[1]);
				pc.z.push_back(targets[i].vertex[2]);
				pc.range.push_back(targets[i].echo_range);
				pc.amplitude.push_back(targets[i].amplitude);
				pc.reflectance.push_back(targets[i].reflectance);
				pc.deviation.push_back(targets[i].deviation);
				pc.return_number.push_back(i+1);
				pc.time.push_back(targets[i].time);
			}	
		}
};

int main(int argc,char** argv)
{
	std::string top_dir = argv[1];
	if(top_dir[top_dir.length()-1] != '/') top_dir = top_dir + "/";
	std::ifstream cfile;
	cfile.open(argv[2]);
	float coordfile[4];
	int no_count = 0;
	if(cfile.is_open())
	{
		while(!cfile.eof())
		{
			cfile >> coordfile[no_count];
			no_count++;
		}
	}
	cfile.close();
	float x_min = coordfile[0]-25;
	float x_max = coordfile[1]+25;
	float y_min = coordfile[2]-25;
	float y_max = coordfile[3]+25;
	float area = atof(argv[3]);
	int length = ceil((x_max-x_min)/sqrt(area)) * ceil((y_max-y_min)/sqrt(area)) + 1;
	float coords[length][4];
	int c = 0;
	int count[length];
	for(float x=x_min;x<x_max;x+=sqrt(area))
	{
		for(float y=y_min;y<y_max;y+=sqrt(area))
		{
			coords[c][0] = x;
			coords[c][1] = x+sqrt(area);
			coords[c][2] = y;
			coords[c][3] = y+sqrt(area);
			count[c] = 0;
			c++;
		}
	}
	coords[length-1][0] = (x_max + x_min) / 2 - 15;
	coords[length-1][1] = (x_max + x_min) / 2 + 15;
	coords[length-1][2] = (y_max + y_min) / 2 - 15;
	coords[length-1][3] = (y_max + y_min) / 2 + 15;
	count[length-1] = 0;
	float deviation_max = atof(argv[4]);
	std::string fname = argv[5];
	std::stringstream ss;
	std::ofstream xyzfiles[length];
	std::string xyznames[length];
	std::string pcdnames[length];
	for(int j=0;j<length;j++)
	{
		if(j == length-1)
		{
			ss.str("");
			ss << fname << ".sample.xyz";
			xyzfiles[j].open(ss.str(),std::ios::binary);
			xyznames[j] = ss.str();
			ss.str("");
			ss << fname << ".sample.pcd";
			pcdnames[j] = ss.str();
		}
		else
		{
			ss.str("");
			ss << fname << "_" << j << ".xyz";
			xyzfiles[j].open(ss.str(),std::ios::binary);
			xyznames[j] = ss.str();
			ss.str("");
			ss << fname << "_" << j << ".pcd";
			pcdnames[j] = ss.str();
		}
	}
	std::vector<std::string> positions;
	DIR *tdir = NULL;
	tdir = opendir (top_dir.c_str());
	struct dirent *tent = NULL;
	while(tent = readdir(tdir)) positions.push_back(tent->d_name);
	closedir(tdir);
	for(int k=0;k<positions.size();k++)
	{
		if(positions[k][0] == 'S' && positions[k][4] == 'P')
		{
			ss.str("");
			ss << top_dir << positions[k];
			std::string position;
			const char* c_position;
			position = ss.str();
			c_position = position.c_str();
			std::vector<std::string> position_contents;
			DIR *pdir = NULL;
			pdir = opendir(c_position);
			struct dirent *pent = NULL;
			while(pent = readdir(pdir)) position_contents.push_back(pent->d_name);
			closedir(pdir);
			std::string rxpname;
			for(int l=0;l<position_contents.size();l++)
			{
				if(position_contents[l][14] == 'r' && position_contents[l][15] == 'x' && position_contents[l][16] == 'p' && position_contents[l].length() == 17 )
				{
					ss.str("");
					ss << top_dir << positions[k] << "/" << position_contents[l];
					rxpname = ss.str();
				}
			}
			ss.str("");
			ss << top_dir << "matrix/" << positions[k][7] << positions[k][8] << positions[k][9] <<".dat";
			std::string matrixname = ss.str();
			std::cout << rxpname << " " << " " << matrixname << std::endl;
			pcloud pc;
			try
			{

				std::shared_ptr<scanlib::basic_rconnection> rc;
				rc = scanlib::basic_rconnection::create(rxpname);
				rc->open();
				scanlib::decoder_rxpmarker dec(rc);
				importer imp(pc);
				scanlib::buffer buf;
				for(dec.get(buf);!dec.eoi();dec.get(buf))
				{
					imp.dispatch(buf.begin(), buf.end());
				}
				rc->close();
			}
			catch(...)
			{
				continue;
			}
			ss.str("");
			if(positions[k][7] == '0' && positions[k][8] == '0') ss << positions[k][9];
			else if(positions[k][7] == '0') ss << positions[k][8] << positions[k][9];
			else ss << positions[k][7] << positions[k][8] << positions[k][9];
			std::string scan_number = ss.str();
			pc.scan_number = atof(scan_number.c_str());
			std::ifstream mfile;
			mfile.open(matrixname);
			int no_count = 0;
			if(mfile.is_open())
			{
				while(!mfile.eof())
				{
					mfile >> pc.matrix[no_count];
					no_count++;
				}
			}
			for(int m=0;m<pc.x.size();m++)
			{
				float X = ((pc.x[m]*pc.matrix[0])+(pc.y[m]*pc.matrix[1])+(pc.z[m]*pc.matrix[2]))+pc.matrix[3];
				float Y = ((pc.x[m]*pc.matrix[4])+(pc.y[m]*pc.matrix[5])+(pc.z[m]*pc.matrix[6]))+pc.matrix[7];
				float Z = ((pc.x[m]*pc.matrix[8])+(pc.y[m]*pc.matrix[9])+(pc.z[m]*pc.matrix[10]))+pc.matrix[11];
				for(int n=0;n<length;n++)
				{
					if(X >= coords[n][0] && X < coords[n][1])
					{
						if(Y >=coords[n][2] && Y < coords[n][3])
						{
							if(pc.deviation[m] <= deviation_max)
							{
								xyzfiles[n].write(reinterpret_cast<const char*>(&X),sizeof(X));
								xyzfiles[n].write(reinterpret_cast<const char*>(&Y),sizeof(Y));
								xyzfiles[n].write(reinterpret_cast<const char*>(&Z),sizeof(Z));
								count[n] += 1;
							}
						}
					}
				}
			}
		}
	}
	for(int p=0;p<length;p++)
	{
		xyzfiles[p].close();
	}
	for(int q=0;q<length;q++)
	{
		std::ofstream headerstream("header.tmp");
		headerstream << "VERSION 0.7" << std::endl << "FIELDS x y z" << std::endl << "SIZE 4 4 4" << std::endl << "TYPE F F F" << std::endl << "COUNT 1 1 1" << std::endl << "WIDTH " << count[q] << std::endl << "HEIGHT 1" << std::endl << "VIEWPOINT 0 0 0 1 0 0 0" << std::endl << "POINTS " << count[q] << std::endl << "DATA binary" << std::endl;
		headerstream.close();
		ss.str("");
		ss << "cat header.tmp " << xyznames[q] << " > " << pcdnames[q] << "; rm header.tmp " << xyznames[q];
		std::string string;
		const char* cc;
		string = ss.str();
		cc = string.c_str();
		system(cc);
	}
	return 0;
}
